### ISSUE TITLE
[Help] <Evaluation>

### ISSUE THREAD

**Issue by charming2992** [THUDM/ChatGLM2-6B#218]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

在使用v2版本评估C-Eval 数据集的时候，在第二轮匹配选项的时候的这一行的具体内容是什么呀？
outputs = model(**inputs, return_last_logit=True)
return_last_logit这个参数的具体作用是？

以及使用v1版本的话这个参数是没有的，如何使用v1版本的评估C-Eval呢

### Expected Behavior

可以解释一下return_last_logit这个具体作用吗

### Steps To Reproduce

python evaluate_ceval.py

### Environment

```markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

**Comment by lilongxian** [comment]
该参数的作用在此：
1， ChatGLMForConditionalGeneration类即chatGLM2模型实例，
return_last_logit作为其forward喂入参数：
def forward(
            self,
            input_ids: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.Tensor] = None,
            attention_mask: Optional[torch.Tensor] = None,
            past_key_values: Optional[Tuple[torch.FloatTensor]] = None,
            inputs_embeds: Optional[torch.Tensor] = None,
            labels: Optional[torch.Tensor] = None,
            use_cache: Optional[bool] = None,
            output_attentions: Optional[bool] = None,
            output_hidden_states: Optional[bool] = None,
            return_dict: Optional[bool] = None,
            return_last_logit: Optional[bool] = False,）

transformer_outputs = self.transformer(
            input_ids=input_ids,。。。)
hidden_states = transformer_outputs[0]
if return_last_logit:
         hidden_states = hidden_states[-1:]

self.transformer 即GLM2 Encoder Net: E[embedding, full_attention_mask, rotary_pos_emb] + GLM2-Encoder[MQA-att-Encoder-loop + LN]
self.transformer网络对输入的编码输出为：
            (
           last_hidden_state=hidden_states,
            past_key_values=presents,
            hidden_states=all_hidden_states,
            attentions=all_self_attentions
          )
网络编码默认hidden_states为layer_27层+LN层的实际编码输出last_hidden_state，代表了整体28各层+LN的编码结果。
all_hidden_states =（layer0_hidden_states ,layer1_hidden_states,.....,layer27_hidden_states ）.存的是tf每层输出但没有LN的hidden_states。
all_self_attentions=（layer0_att_hidden_states ,layer1_att_states,.....,layer27_att_states）指所有28各层的MQA-ATT输出的序列。
从上面代码可以看到return_last_logit作用仅仅是提示是否指定GLM网络编码输出all_self_attentions。

GLM1/2中，selfATT机制单层输出编码为：hidden_states【真实编码输出张量】,  kv_cache【MQA-ATT的共享k、v，用于use_cache时后加速运行模型（仅在使用梯度检查点训练时不需要），主要用在验证、推理加速】。

在GLM中：
 if not return_dict:
        return tuple(v for v in [hidden_states, presents, all_hidden_states, all_self_attentions] if v is not None)

return BaseModelOutputWithPast(
            last_hidden_state=hidden_states,
            past_key_values=presents,
            hidden_states=all_hidden_states,
            attentions=all_self_attentions,
        )
上面代码可以看出：
当 return_dict 为False或None且return_last_logit=Ture 时，self.transformer返回（hidden_states, presents, all_hidden_states），这个时候， GLM网络输出的 hidden_states 就是all_hidden_states ==（layer0_hidden_states ,layer1_hidden_states,.....,layer27_hidden_states ）而不是all_self_attentions。
return_dict 为Ture且return_last_logit=False时，self.transformer返回（hidden_states, presents, all_hidden_states，all_self_attentions），GLM网络输出的 hidden_states 就是all_self_attentions。

在hf transformers的class PretrainedConfig(PushToHubMixin)类中，
self.return_dict = kwargs.pop("return_dict", True)
        self.output_hidden_states = kwargs.pop("output_hidden_states", False)
        self.output_attentions = kwargs.pop("output_attentions", False)
        self.torchscript = kwargs.pop("torchscript", False)  # Only used by PyTorch models
 def use_return_dict(self) -> bool:
        """
        `bool`: Whether or not return [`~utils.ModelOutput`] instead of tuples.
        """
        # If torchscript is set, force `return_dict=False` to avoid jit errors
        return self.return_dict and not self.torchscript
以及GLM： return_dict = return_dict if return_dict is not None else self.config.use_return_dict  并且没有另外专门设置该参数
可看出：return_dict默认为True, 所以上述return_last_logit=Ture 代表获取GLM的all_self_attentions序列

**Comment by lilongxian** [comment]
你也可以看这里，这个函数很重要：
    def prepare_inputs_for_generation(
            self,
            input_ids: torch.LongTensor,
            past_key_values: Optional[torch.Tensor] = None,
            attention_mask: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.Tensor] = None,
            is_first_forward: bool = True,
            **kwargs
    ) -> dict:
        # only last token for input_ids if past is not None
        if position_ids is None:
            position_ids = self.get_position_ids(input_ids, device=input_ids.device)
        if not is_first_forward:
            position_ids = position_ids[..., -1:]
            input_ids = input_ids[:, -1:]

        return {
            "input_ids": input_ids,
            "past_key_values": past_key_values,
            "position_ids": position_ids,
            "attention_mask": attention_mask,
            "return_last_logit": True 
        }

其中"return_last_logit" 这个参数在验证和预测中用于从网络编码输出logist中抽取最后那个token的向量表征, 以在beam_search中搜索最佳的token概率，即生成下一个token!  所以并不只是在验证脚本中出现。

**Comment by duzx16** [comment]
这个参数的作用是只返回最后一个 token 的 logit，用来节省计算量的
